randomized_svd: More efficient SVD computation

matospiso opened the pull request:
Efficiency improvement in randomized_svd: compute SVD of a tall and thin matrix instead of a wide matrix

<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Despite the code stating 
> compute the SVD on the **_thin_** matrix: (k + p) wide

the matrix B has (k + p) _rows_, not columns; it is wide. For a wide matrix (`shape[1] > shape[0]`) it is more efficient to compute the SVD of its transpose and then transpose the factors.

Complexity of SVD (assuming $m \times n$ matrix) is $O(mn^2)$, so if $m < n$ it is cheaper to perform the operation on the matrix transpose.

#### Any other comments?
See e.g. https://mathoverflow.net/a/221216 for a brief complexity analysis.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

matospiso committed:
Update extmath.py

Efficiency improvement in randomized_svd: compute SVD of a tall and thin matrix instead of a wide matrix

matospiso committed:
black

ogrisel commented:
Could you please run some benchmarks on this branch and compare the same benchmarks run on `main`?

The following scripts might be directly reusable for this.

- `benchmarks/bench_plot_randomized_svd.py`
- `benchmarks/bench_pca_solvers.py`

Please post the results either as comments or directly by editing the description of the PR.

ogrisel commented on sklearn/utils/extmath.py:
```diff
@@ -530,21 +530,26 @@ def randomized_svd(
-    # compute the SVD on the thin matrix: (k + p) wide
+    # Typically M.shape[1] > k + p, therefore we
+    # compute the SVD of the tall and thin matrix B.T with shape (M.shape[1], k + p)
+    # instead of less efficient SVD of (wide) matrix B.
```
For reference, I ran the following quick experiment to check the empirical validity of the above comment:

```python
>>> import numpy as np
>>> a = np.random.normal(size=(10_000, 10))  # tall and thin
>>> %timeit np.linalg.svd(a, full_matrices=False)
1.67 ms ± 1.87 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit np.linalg.svd(a.T, full_matrices=False)
3.01 ms ± 7.39 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> a.flags
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

Again, with a transposed memory layout:

```python
>>> b = np.random.normal(size=(10, 10_000))
>>> %timeit np.linalg.svd(b.T, full_matrices=False)  # tall and thin
1.61 ms ± 2.3 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
>>> %timeit np.linalg.svd(b, full_matrices=False)
2.99 ms ± 7.33 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
>>> b.flags
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

Same conclusion: tall and thin is always the fastest, whatever the memory layout.

ogrisel replied:
I also tried with `scipy.linalg.svd(a, full_matrices=False, lapack_driver="gesvd")` and `scipy.linalg.svd(a, full_matrices=False, lapack_driver="gesdd")` instead of NumPy and got similar results.

matospiso commented:
## `benchmarks/bench_plot_randomized_svd.py`
### Results
In all sections, top=`main`, bottom=`patch-1`
#### `low rank matrix` (500 x 10 000)
![Figure_1](https://github.com/user-attachments/assets/e5d5f5c6-01ae-4759-8d07-5bff99785551)
![Figure_1](https://github.com/user-attachments/assets/16e7f5fd-f59c-4fdc-b925-46b615a89e67)

#### `lfw_people`
skipped manually

#### `olivetti_faces` (400 x 4 096)
![Figure_2](https://github.com/user-attachments/assets/485370d2-530e-41c6-aa83-1abe6c428d33)
![Figure_2](https://github.com/user-attachments/assets/787bfb7f-8987-4cce-a8b3-17a1ff42b166)

#### `20newsgroups` (11 314 x 100 000)
![Figure_3](https://github.com/user-attachments/assets/319fbfae-0dce-459c-9d99-f222decc765c)
![Figure_3](https://github.com/user-attachments/assets/20416fbc-715a-44aa-846f-0bccc1f4de7b)

#### `mnist_784` (70 000 x 784)
![Figure_4](https://github.com/user-attachments/assets/de666151-866c-4160-8250-57efc7a5f3ff)
![Figure_4](https://github.com/user-attachments/assets/030af13b-ceb7-49f1-9eb1-d4d21d67ac2a)

#### `CIFAR`
skipped (file folder not found)

#### `a3a` (32 561 x 123)
![Figure_5](https://github.com/user-attachments/assets/589a358c-7d5d-40f7-b43e-50d09047ec37)
![Figure_5](https://github.com/user-attachments/assets/2402689d-6056-4cda-8578-2f1f91e8169e)

#### `SVHN`
skipped (file folder not found)

#### `uncorrelated matrix` (500 x 10 000)
![Figure_6](https://github.com/user-attachments/assets/a60302e6-9713-4f32-bf57-98f750dfbe10)
![Figure_6](https://github.com/user-attachments/assets/b033a7f4-36c0-4430-bd8e-a7ddcae8a8da)

#### `big sparse matrix` (1 000 000 x 10 000)
no plot generated

#### `rcv1` (804 414 x 47 236)
no plot generated

### Comments
- `scikit-learn==1.5.2`, `main` and `patch-1` (my branch) all crash in `benchmarks/bench_plot_randomized_svd.py` on
```
Getting dataset: lfw_people
>>>>>> Benching sklearn and fbpca on lfw_people 13233 x 2914
... [several computations] ...
n_iter = 5 on sklearn - none
... running SVD ...
Traceback (most recent call last):
  File "/Users/mato/codes/scikit-learn/benchmarks/bench_plot_randomized_svd.py", line 514, in <module>
    bench_a(
  File "/Users/mato/codes/scikit-learn/benchmarks/bench_plot_randomized_svd.py", line 356, in bench_a
    U, s, V, time = svd_timing(
                    ^^^^^^^^^^^
  File "/Users/mato/codes/scikit-learn/benchmarks/bench_plot_randomized_svd.py", line 282, in svd_timing
    U, mu, V = randomized_svd(
               ^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/sklearn/lib/python3.11/site-packages/sklearn/utils/_param_validation.py", line 213, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/sklearn/lib/python3.11/site-packages/sklearn/utils/extmath.py", line 539, in randomized_svd
    Uhat, s, Vt = linalg.svd(
                  ^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/sklearn/lib/python3.11/site-packages/scipy/linalg/_decomp_svd.py", line 108, in svd
    a1 = _asarray_validated(a, check_finite=check_finite)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/sklearn/lib/python3.11/site-packages/scipy/_lib/_util.py", line 321, in _asarray_validated
    a = toarray(a)
        ^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/sklearn/lib/python3.11/site-packages/numpy/lib/_function_base_impl.py", line 656, in asarray_chkfinite
    raise ValueError(
ValueError: array must not contain infs or NaNs
```
I'm pretty sure this is overflow, and likely expected behavior. The benchmark shouldn't crash though, however, I'm not sure how the fix should look, so I commented out the failing dataset `lfw_people`.

- `mnist_784` crashes with 
```
Getting dataset: mnist_784
 >>>>>> Benching sklearn and fbpca on mnist_784 70000 x 784
n_iter = 0 on sklearn - none
... running SVD ...
Traceback (most recent call last):
  File "/Users/mato/codes/scikit-learn/benchmarks/bench_plot_randomized_svd.py", line 514, in <module>
    bench_a(
  File "/Users/mato/codes/scikit-learn/benchmarks/bench_plot_randomized_svd.py", line 356, in bench_a
    U, s, V, time = svd_timing(
                    ^^^^^^^^^^^
  File "/Users/mato/codes/scikit-learn/benchmarks/bench_plot_randomized_svd.py", line 282, in svd_timing
    U, mu, V = randomized_svd(
               ^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniconda/base/envs/sklearn/lib/python3.11/site-packages/sklearn/utils/_param_validation.py", line 206, in wrapper
    validate_parameter_constraints(
  File "/opt/homebrew/Caskroom/miniconda/base/envs/sklearn/lib/python3.11/site-packages/sklearn/utils/_param_validation.py", line 98, in validate_parameter_constraints
    raise InvalidParameterError(
sklearn.utils._param_validation.InvalidParameterError: The 'M' parameter of randomized_svd must be an instance of 'numpy.ndarray' or a sparse matrix. Got        pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  pixel8  ...  pixel777  pixel778  pixel779  pixel780  pixel781  pixel782  pixel783  pixel784
0           0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
1           0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
2           0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
3           0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
4           0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
...       ...     ...     ...     ...     ...     ...     ...     ...  ...       ...       ...       ...       ...       ...       ...       ...       ...
69995       0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
69996       0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
69997       0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
69998       0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0
69999       0       0       0       0       0       0       0       0  ...         0         0         0         0         0         0         0         0

[70000 rows x 784 columns] instead.
```
so I replaced `X = fetch_openml(dataset_name).data` (line 195) with
```
X = fetch_openml(dataset_name).data
    if isinstance(X, pd.DataFrame):
    X = X.to_numpy()
```
We should add this change to fix the benchmark as part of this PR.

- it would be nice to see the plots for the last two matrices (+ their transpositions) as well (or at least log the runtime)

matospiso commented:
@ogrisel one more "large-scale" benchmark:
```python
Python 3.11.10 (main, Oct  3 2024, 02:26:51) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.29.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np

In [2]: a = np.random.normal(size=(1_000_000, 256))

In [3]: %timeit np.linalg.svd(a, full_matrices=False)
9.93 s ± 434 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [4]: %timeit np.linalg.svd(a.T, full_matrices=False)
21.9 s ± 1.29 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
here it starts to really matter if the matrix is thin or wide :)

ogrisel commented:
From the results in https://github.com/scikit-learn/scikit-learn/pull/30256#issuecomment-2482801981 we globally observe a small performance improvement on most datasets.

The only exception is the `a3a` where it seems that there is ~10% slowdown with the new branch. It would be good to understand what's going on here.

I would also be interested in the impact of this PR on the results of running `bench_pca_solvers.py` compared to `main`.

ogrisel commented:
> here it starts to really matter if the matrix is thin or wide :)

I reproduce the ~2x speed-up on the `(1_000_000, 256)` data. However, that speed-up does not improve when making the matrix even thinner:

```python
In [1]: import numpy as np

In [2]: a = np.random.normal(size=(int(1e6), 256))

In [3]: %time _ = np.linalg.svd(a, full_matrices=False)
CPU times: user 1min 45s, sys: 2.94 s, total: 1min 48s
Wall time: 17.1 s

In [4]: %time _ = np.linalg.svd(a.T, full_matrices=False)
CPU times: user 3min, sys: 2.73 s, total: 3min 3s
Wall time: 35.1 s

In [5]: 35.1 / 17.1
Out[5]: 2.052631578947368

In [6]: a = np.random.normal(size=(int(1e7), 26))

In [7]: %time _ = np.linalg.svd(a, full_matrices=False)
CPU times: user 27.8 s, sys: 6.34 s, total: 34.2 s
Wall time: 9.31 s

In [8]: %time _ = np.linalg.svd(a.T, full_matrices=False)
CPU times: user 59.3 s, sys: 2.34 s, total: 1min 1s
Wall time: 13.8 s

In [9]: 13.8 / 9.31
Out[9]: 1.4822771213748658
```

matospiso commented:
> The only exception is the `a3a` where it seems that there is ~10% slowdown with the new branch. It would be good to understand what's going on here.

I noticed it, too, but the matrix is quite small + it's _thin_. It might be the case that transposition here actually hurts, and the correct solution would be to check if `A.shape[0] >= A.shape[1]` and transpose only if this condition does not hold. 

I'll first rerun `a3a` a couple of times to see if I observe the same slowdown consistently, and if that's the case, I'll implement the conditional transposition and post the benchmarks again.

matospiso commented:
> > here it starts to really matter if the matrix is thin or wide :)
> 
> I reproduce the ~2x speed-up on the `(1_000_000, 256)` data. However, that speed-up does not improve when making the matrix even thinner:
> 
> ```python
> In [1]: import numpy as np
> 
> In [2]: a = np.random.normal(size=(int(1e6), 256))
> 
> In [3]: %time _ = np.linalg.svd(a, full_matrices=False)
> CPU times: user 1min 45s, sys: 2.94 s, total: 1min 48s
> Wall time: 17.1 s
> 
> In [4]: %time _ = np.linalg.svd(a.T, full_matrices=False)
> CPU times: user 3min, sys: 2.73 s, total: 3min 3s
> Wall time: 35.1 s
> 
> In [5]: 35.1 / 17.1
> Out[5]: 2.052631578947368
> 
> In [6]: a = np.random.normal(size=(int(1e7), 26))
> 
> In [7]: %time _ = np.linalg.svd(a, full_matrices=False)
> CPU times: user 27.8 s, sys: 6.34 s, total: 34.2 s
> Wall time: 9.31 s
> 
> In [8]: %time _ = np.linalg.svd(a.T, full_matrices=False)
> CPU times: user 59.3 s, sys: 2.34 s, total: 1min 1s
> Wall time: 13.8 s
> 
> In [9]: 13.8 / 9.31
> Out[9]: 1.4822771213748658
> ```
Here it's better to look for speedup w.r.t. CPU time as the `gesvd`/`gesdd` LAPACK functions for computing SVD are very well parallelizable and hence CPU correlates with the total FLOPs performed. In CPU time, the speedup grows as the matrix gets thinner.

ogrisel commented:
> We should add this change to fix the benchmark as part of this PR.

+1.

matospiso commented:
I ran the `a3a` part of `benchmarks/bench_plot_randomized_svd.py` two more times on both `main` and `patch-1`, and this time the times look pretty much identical.
### `main`
Run 1
![main-run-1](https://github.com/user-attachments/assets/3fbe9d30-46a4-417c-8ee5-ce4234f19952)
Run 2
![main-run-2](https://github.com/user-attachments/assets/865ab313-e965-4d49-9dab-67df7b6e3fb1)
### `patch-1`
Run 1
![patch-1-run-1](https://github.com/user-attachments/assets/670a71c7-438d-45f9-834a-e60e4b097fb1)
Run 2
![patch-1-run-2](https://github.com/user-attachments/assets/8a144322-46a7-48ab-97cf-2d53c5e7157e)

Also, in `patch-1` run 1 you can observe a random 20ms slowdown in one of the simplest configuration. I believe this is unrelated to the actual change (cause is likely process-level scheduling)

matospiso committed:
fix crashing mnist dataset (loading error), skip crashing lfw_people dataset (overflow)

matospiso committed:
transpose B only if it's wide; refactor to improve readability

matospiso commented:
@ogrisel 
- added a check if B is thin or not. It'll transpose B only if it's wide
- fixed loading of MNIST in `bench_plot_randomized_svd.py`
- commented out the `lfw_people` dataset in the benchmark and added a TODO to avoid the overflow or handle it gracefully

ogrisel reviewed (commented):
Thanks very much for the follow-up.

Could you identity a case where this change leads to a large improvement for the `randomized_svd` or `PCA` call as a whole?

While this PR can lead to a 2x speed-up for the final SVD step, I have the impression that the cost of `randomized_range_finder` always dominates and makes optimizing the final SVD call in `randomized_svd` a bit irrelevant.

That being said, I am not opposed to merging this PR because the extra-complexity is limited.

Also do you have good reference for the complexity analysis of the LAPACK SVD solvers? I could not find any after a quick googling and the docstring of numpy and scipy do not mention the complexity nor the fact that it is dependent on the transposition of the input data.

ogrisel commented:
And we would also need a changelog entry for this change. But given the fact that it barely impacts the runtime of `randomized_svd` in cases where we did the measurement, I am not sure how to present this.
